activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/ProcessDeletedEventHandler.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017-2020 Alfresco Software, Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.activiti.cloud.services.query.events.handlers;
+
+import java.util.Optional;
+import java.util.Set;
+import javax.persistence.EntityManager;
+import org.activiti.api.process.model.ProcessInstance;
+import org.activiti.api.process.model.ProcessInstance.ProcessInstanceStatus;
+import org.activiti.cloud.api.model.shared.events.CloudRuntimeEvent;
+import org.activiti.cloud.api.process.model.events.CloudProcessDeletedEvent;
+import org.activiti.cloud.api.process.model.events.ExtendedCloudProcessRuntimeEvent.ExtendedCloudProcessRuntimeEvents;
+import org.activiti.cloud.services.query.model.ProcessInstanceEntity;
+import org.activiti.cloud.services.query.model.QueryException;
+
+public class ProcessDeletedEventHandler implements QueryEventHandler {
+
+  protected final String INVALID_PROCESS_INSTANCE_STATE = "Process Instance %s is not in a valid state: %s";
+
+  private Set<ProcessInstanceStatus> ALLOWED_STATUS = Set.of(ProcessInstanceStatus.CANCELLED, ProcessInstanceStatus.COMPLETED);
+
+  private final EntityManager entityManager;
+
+  public ProcessDeletedEventHandler(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
+
+  @Override
+  public void handle(CloudRuntimeEvent<?, ?> event) {
+    CloudProcessDeletedEvent deletedEvent = (CloudProcessDeletedEvent) event;
+
+    ProcessInstance eventProcessInstance = deletedEvent.getEntity();
+
+    ProcessInstanceEntity processInstanceEntity = Optional.ofNullable(entityManager.find(ProcessInstanceEntity.class,
+            eventProcessInstance.getId()))
+        .orElseThrow(() -> new QueryException("Unable to find process instance with the given id: " + eventProcessInstance.getId()));
+
+    if (ALLOWED_STATUS.contains(processInstanceEntity.getStatus())) {
+      entityManager.remove(processInstanceEntity);
```

Comment:
Actually, I think this will be more complex than that. IIRC, we have intentionally created the database schema for query service without adding the foreign key contraints, meaning that data related to the process instance will not get automatically deleted (user tasks, service tasks, etc). I think it's worth to turn `ProcessDeletedEventHandlerTest ` into an integration test where you use `ProcessInstanceRepository`, `TaskRepository`, `ServiceTaskRepository`, etc  to insert data and check if they are deleted in the end.

Reply:
Done, thank you for noticing.
The same behavior is implemented in `org.activiti.cloud.services.query.rest.ProcessInstanceDeleteController.deleteProcessInstances`.